src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -0,0 +1,554 @@
+
+    :param ast: reference into the fparser2 parse tree corresponding to \
+                this node.
+    :type ast: sub-class of :py:class:`fparser.two.Fortran2003.Base`
+    :param children: a list of child nodes for this node. These will be made \
+                     children of the child Schedule of this PSyDataNode.
+    :type children: list of :py:class:`psyclone.psyir.nodes.Node` \
+                    or derived classes
+    :param parent: the parent of this node in the PSyIR.
+    :type parent: :py:class:`psyclone.psyir.nodes.Node`
+    :param (str,str) name: an optional name to use for this PSyDataNode, \
+        provided as a 2-tuple containing a module name followed by a \
+        local name. The pair of strings should uniquely identify a\
+        region unless aggregate information is required and supported
+        by the runtime library linked in.
+
+    '''
+    # PSyData interface Fortran module
+    fortran_module = "psy_data_mod"
+    # The symbols we import from the PSyData Fortran module
+    symbols = ["PSyDataType"]
+    # The use statement that will be inserted. Any use of a module of the
+    # same name that doesn't match this will result in a NotImplementedError
+    # at code-generation time.
+    use_stmt = "use psy_data_mod, only: " + ", ".join(symbols)
+
+    # Root of the name to use for variables associated with PSyData regions
+    psy_data_var = "psy_data"
+
+    def __init__(self, ast=None, children=None, parent=None, name=None):
+        # TODO: #415 Support different classes of PSyData calls.
+
+        # Store the name of the PSyData variable that is used for this
+        # PSyDataNode. This allows the variable name to be shown in str
+        # (and also, calling create_name in gen() would result in the name
+        # being changed every time gen() is called).
+        from psyclone.psyGen import NameSpaceFactory
+        self._var_name = NameSpaceFactory().create().create_name("psy_data")
+
+        if children:
+            # We need to store the position of the original children,
+            # i.e. before they are added to a schedule
+            node_parent = children[0].parent
+            node_position = children[0].position
+
+        # A PSyData node always contains a Schedule
+        sched = self._insert_schedule(children)
+        super(PSyDataNode, self).__init__(ast=ast, children=[sched],
+                                          parent=parent)
+
+        if children:
+            # Correct the parent's list of children. Use a slice of the list
+            # of nodes so that we're looping over a local copy of the list.
+            # Otherwise things get confused when we remove children from
+            # the list.
+            for child in children[:]:
+                # Remove child from the parent's list of children
+                node_parent.children.remove(child)
+
+            # Add this node as a child of the parent
```

Comment:
Don't we want to add this node to the parent irrespective of whether `children` were supplied? Please check that this aspect is covered by the test suite.

Reply:
While I think that in reality we always will have children (afaik the transforms would reject 'None' as children, and an empty list would actually cause a crash:
```
if isinstance(nodes, list) and isinstance(nodes[0], Node):
```
) I have fixed this, and added a long test for various combinations of parent and children.